src/meltano/core/plugin_invoker.py

```diff
@@ -321,10 +324,63 @@ def env(self):
         Returns:
             Dictionary of environment variables.
         """
+        project_settings_service = ProjectSettingsService(
+            self.project, config_service=self.plugins_service.config_service
+        )
+        with self.settings_service.feature_flag(
+            FeatureFlags.STRICT_ENV_VAR_MODE, raise_error=False
+        ) as strict_env_var_mode:
+
+            # Expand root env w/ os.environ
+            expanded_project_env = expand_env_vars(
+                project_settings_service.env,
+                os.environ,
+                raise_if_missing=strict_env_var_mode,
+            )
+            expanded_project_env.update(
+                expand_env_vars(
+                    self.settings_service.project.dotenv_env,
+                    os.environ,
+                    raise_if_missing=strict_env_var_mode,
+                )
+            )
+            # Expand active env w/ expanded root env
+            expanded_active_env = (
+                expand_env_vars(
+                    self.settings_service.project.active_environment.env,
+                    expanded_project_env,
+                    raise_if_missing=strict_env_var_mode,
+                )
+                if self.settings_service.project.active_environment
+                else {}
+            )
+
+            # Expand root plugin env w/ expanded active env
+            expanded_root_plugin_env = expand_env_vars(
+                self.settings_service.plugin.env,
+                expanded_active_env,
+                raise_if_missing=strict_env_var_mode,
+            )
+
+            # Expand active env plugin env w/ expanded root plugin env
+            expanded_active_env_plugin_env = (
+                expand_env_vars(
+                    self.settings_service.environment_plugin_config.env,
+                    expanded_root_plugin_env,
+                    raise_if_missing=strict_env_var_mode,
+                )
+                if self.settings_service.environment_plugin_config
+                else {}
+            )
```

Comment:
Nice! It's very clear what's going on here and which variables are used to expand the next level down 👍

Reply:
Big +1